verifiers/internal/gcb/provenance.go

```diff
@@ -0,0 +1,308 @@
+		return "", fmt.Errorf("%w: '%s' is not a string", serrors.ErrorInvalidDssePayload, key)
+	}
+	return ts, nil
+}
+
+func (self *GCBProvenance) VerifySubjectDigest(expectedHash string) error {
+	if err := self.isVerified(); err != nil {
+		return err
+	}
+
+	statement := self.verifiedIntotoStatementStruct
+	for _, subject := range statement.StatementHeader.Subject {
+		digestSet := subject.Digest
+		hash, exists := digestSet["sha256"]
+		if !exists {
+			return fmt.Errorf("%w: %s", serrors.ErrorInvalidDssePayload, "no sha256 subject digest")
+		}
+
+		if hash == expectedHash {
+			return nil
+		}
+	}
+
+	return fmt.Errorf("expected hash '%s' not found: %w", expectedHash, serrors.ErrorMismatchHash)
+}
+
+// Verify source URI in provenance statement.
+func (self *GCBProvenance) VerifySourceURI(expectedSourceURI string) error {
+	if err := self.isVerified(); err != nil {
+		return err
+	}
+
+	statement := self.verifiedIntotoStatementStruct
+	materials := statement.Predicate.Materials
+	if len(materials) == 0 {
+		return fmt.Errorf("%w: no materials", serrors.ErrorInvalidDssePayload)
+	}
+	uri := materials[0].URI
+	if !strings.HasPrefix(expectedSourceURI, "https://") {
+		expectedSourceURI = "https://" + expectedSourceURI
+	}
+	if !strings.HasPrefix(uri, expectedSourceURI+"/commit/") {
+		return fmt.Errorf("%w: expected '%s', got '%s'",
+			serrors.ErrorMismatchSource, expectedSourceURI, uri)
+	}
+
+	return nil
+}
+
+func (self *GCBProvenance) VerifyBranch(branch string) error {
+	return fmt.Errorf("%w: GCB branch verification", serrors.ErrorNotSupported)
+}
+
+func (self *GCBProvenance) VerifyTag(tag string) error {
+	return fmt.Errorf("%w: GCB tag verification", serrors.ErrorNotSupported)
+}
+
+func (self *GCBProvenance) VerifyVersionedTag(tag string) error {
+	return fmt.Errorf("%w: GCB versioned-tag verification", serrors.ErrorNotSupported)
+}
+
+func (self *GCBProvenance) VerifySignature() error {
+	if len(self.gcloudProv.ProvenanceSummary.Provenance) == 0 {
+		return fmt.Errorf("%w: no provenance found", serrors.ErrorInvalidDssePayload)
+	}
+	// Assume a single provenance in the array.
+	prov := self.gcloudProv.ProvenanceSummary.Provenance[0]
+
+	// Verify the envelope type. It should be an intoto type.
+	if prov.Envelope.PayloadType != intoto.PayloadType {
+		return fmt.Errorf("%w: expected payload type '%s', got %s",
+			serrors.ErrorInvalidDssePayload, intoto.PayloadType, prov.Envelope.PayloadType)
+	}
+
+	payload, err := payloadFromEnvelope(&prov.Envelope)
+	if err != nil {
+		return err
+	}
+
+	payloadHash := sha256.Sum256(payload)
+
+	var errs []error
+	regex := regexp.MustCompile(`^projects\/verified-builder\/locations\/(.*)\/keyRings\/attestor\/cryptoKeys\/builtByGCB\/cryptoKeyVersions\/1$`)
+
+	for _, sig := range prov.Envelope.Signatures {
+		match := regex.FindStringSubmatch(sig.KeyID)
+		if len(match) == 2 {
+			// Create a public key instance for this region.
+			region := match[1]
+			pubKey, err := keys.PublicKeyNew(region)
```

Comment:
maybe this should return a dsse verifier? that way you don't have to implement the payload hashing or signature decoding.

Reply:
I had to replace 2 characters before base64 decoding: Is this standard in dsse? It did not seem to be https://github.com/secure-systems-lab/dsse/blob/master/envelope.md

Reply:
@ianlewis suggested using the base64 URL decoding instead. If that's enough, I'll use this. @asraa would that allow me to use the "dsse verifier"? Is it defined on their dsse repo?

Reply:
Updated to use URL decoding.

Reply:
Offline discussion: I can update to use dsse verifier uniformly over all the verifier types (GHA, GCB here)